### PR TITLE
Update JS API for exnref

### PR DESCRIPTION
**Pull request by dschuff**
This change updates exception object allocation, initialization, and construction for exnref;
as well as dealing with exception propagation from invoking exported functions; throwing
exceptions from host functions into wasm; and wrapping and unwrapping JS exceptions as they 
propagate into and out of wasm.

It depends on the embedding interfaces added in #268.

### CODE REVIEW

File: document/js-api/index.bs [hunk 1]
```diff
@@ -1294,22 +1307,22 @@ constructor steps are:
 1. Let |wasmPayload| be « ».
 1. [=list/iterate|For each=] |value| and |resultType| of |payload| and |types|, paired linearly,
     1. [=list/Append=] ? [=ToWebAssemblyValue=](|value|, |resultType|) to |wasmPayload|.
-1. Set **this**.\[[Type]] to |exceptionTag|.\[[Address]].
-1. Set **this**.\[[Payload]] to |wasmPayload|.
+1. Let (|store|, |exceptionAddr|) be [=exn_alloc=](|store|, |exceptionTag|.\[[Address]], |wasmPayload|)
```
Comment (rossberg) [inline document/js-api/index.bs:1316]: This is missing a consecutive step like
```
1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
```

File: document/js-api/index.bs [hunk 2]
```diff
@@ -1294,22 +1307,22 @@ constructor steps are:
...
+1. [=initialize an Exception object|Initialize=] **this** from |exceptionAddr|.
```
Comment (rossberg) [inline document/js-api/index.bs:1318]: I'm a bit confused where `**this**` is coming from, shouldn't it be allocated first? This is describing the call to `new`, so the object does not pre-exist.

Comment (Ms2ger) [reply document/js-api/index.bs:1318]: As it happens, it does pre-exist, per steps 1.8-1.9 at https://webidl.spec.whatwg.org/#create-an-interface-object

File: document/js-api/index.bs [hunk 3]
```diff
@@ -760,7 +766,7 @@ Each {{Table}} object has a \[[Table]] internal slot, which is a [=table address
     The <dfn constructor for="Table">Table(|descriptor|, |value|)</dfn> constructor, when invoked, performs the following steps:
     1. Let |elementType| be [=ToValueType=](|descriptor|["element"]).
     1. If |elementType| is not a [=reftype=],
-        1. [=Throw=] a {{TypeError}} exception.
+        1. Throw a {{TypeError}} exception.
```
Comment (Ms2ger) [inline document/js-api/index.bs:773]: Why this change?

File: document/js-api/index.bs [hunk 4]
```diff
@@ -1264,20 +1266,31 @@ interface Exception {
 An {{Exception}} value represents an exception.
 
 <div algorithm>
+To <dfn>initialize an Exception object</dfn> |exn| from an [=Exception address=] |exnAddress|, perform the following steps:
```
Comment (Ms2ger) [inline document/js-api/index.bs]: Nit:
```suggestion
<div algorithm>

To <dfn>initialize an Exception object</dfn> |exn| from an [=Exception address=] |exnAddress|, perform the following steps:
```

File: document/js-api/index.bs [hunk 5]
```diff
@@ -1294,22 +1307,22 @@ constructor steps are:
...
 1. [=list/iterate|For each=] |value| and |resultType| of |payload| and |types|, paired linearly,
     1. [=list/Append=] ? [=ToWebAssemblyValue=](|value|, |resultType|) to |wasmPayload|.
-1. Set **this**.\[[Type]] to |exceptionTag|.\[[Address]].
-1. Set **this**.\[[Payload]] to |wasmPayload|.
+1. Let (|store|, |exceptionAddr|) be [=exn_alloc=](|store|, |exceptionTag|.\[[Address]], |wasmPayload|)
+1. [=initialize an Exception object|Initialize=] **this** from |exceptionAddr|.
 1. If |options|["traceStack"] is true,
     1. Set **this**.\[[Stack]] to either a {{DOMString}} representation of the current call stack or undefined.
-1. Otherwise,
-    1. Set **this**.\[[Stack]] to undefined.
```
Comment (Ms2ger) [inline document/js-api/index.bs:1302]: This seems to leave [[Stack]] uninitialized? Does the rest of the spec deal with that?

**Comment by eqrion** [comment]
I don't see this in this PR or the overview, but is it intended for `exn` to be passable to/from JS through params/results, or be like `v128` and cause TypeError's?

File: document/js-api/index.bs (hunk 3 above)
Comment (dschuff) [reply document/js-api/index.bs:773]: Using this annotation around Throw makes it point to the definition of throwing a wasm exception (https://webassembly.github.io/exception-handling/js-api/index.html#webassembly-throw) but this step is actually just meant to throw a regular JS exception. Note how the other places where this spec says to throw JS exceptions for error conditions just use the word "throw" with no links.

File: document/js-api/index.bs (hunk 5 above)
Comment (dschuff) [reply document/js-api/index.bs:1302]: This constructor now calls "initialize an Exception object" which sets [[Stack]] to be `undefined`.

**Comment by dschuff** [comment]
> I don't see this in this PR or the overview, but is it intended for `exn` to be passable to/from JS through params/results, or be like `v128` and cause TypeError's?

When Andreas updated the core spec he also did a minimal update to this spec that made it the latter (see e.g. [here](https://github.com/WebAssembly/exception-handling/blob/main/document/js-api/index.bs#L1003) or in the getters/setters for table elements and globals).
In this PR when exceptions are thrown or propagate from wasm into JS, they end up wrapped in a `WebAssembly.Exception` object (that's line 1023 in this patch) just as they are in phase 3 exceptions. I'm not sure what a concrete use case would be for passing/returning but it seems slightly weird and asymmetric that you can get any exception instance out to JS by throwing it, but not by returning it.

**Comment by rossberg** [comment]
It is intentional. I can't find the original discussion right now, but being able to pass exception packages as first-class values to JS would have nasty implications, such as:

- Instead of being able to just do reference counting on exnrefs internally, it would require engines to do either _cross host-wasm GC_  and hence effectively create a dependency on GC again, or rely on host code correctly participating in RC, which could easily corrupt the Wasm engine when that's done wrong.
- The host could pass it back as an externref, which implies that exnref becomes a subset of externref, which in turn means that GC/RC for exnrefs would have to [include externrefs](https://github.com/WebAssembly/exception-handling/issues/282#issuecomment-1726042705), with very unclear implications for non-JS embedding.
- For JS embeddings, it would likely [expose or constrain implementation details](https://github.com/WebAssembly/exception-handling/pull/269#issuecomment-2007168832) of the handling of the special JavaScript exception tag, because JS code could compare the identity of a thrown JS value and an exnref representing its catch.

Conceptually, its better to think of and exnref as not the exception instance being thrown, but a "pair" of an exception instance that's been caught _plus_ abstract context information for rethrowing it that a Wasm engine might need internally. As such it is a different kind of object from the exception instance and entirely Wasm-internal.

**Comment by eqrion** [comment]
Okay, that was my understanding as well, but wanted to make sure. SpiderMonkey throws a TypeError in ToJSValue.

**Comment by dschuff** [comment]
I think this PR is now reasonably correct and self-contained according to the description in OP. I think it correctly (although may not yet completely) implements what we discussed in #282. (I had forgotten about that thread, we can move further discussion about the desired value semantics back there).
So I think it probably makes sense to finish this review and land it, to keep PRs of a reasonable size.

File: document/js-api/index.bs [hunk 6]
```diff
@@ -1336,18 +1355,31 @@ The <dfn attribute for="Exception">stack</dfn> getter steps are:
...
+    1. return |tagAddress|.
+
+</div>
 
 <div algorithm>
 
-To <dfn for=WebAssembly>throw</dfn> with a [=tag address=] |type|, a matching [=list=] of WebAssembly values |payload|, and an [=externref=] |opaqueData|, perform the following steps:
+To <dfn for=WebAssembly>throw</dfn> with a [=tag address=] |type|, a matching [=list=] of WebAssembly values |payload|, and an [=exception address=] |address|, perform the following steps:
 
 1. Unwind the stack until reaching the *catching try block* given |type|.
```
Comment (dschuff) [inline document/js-api/index.bs]: This is the main bit that still needs to be updated, as there is currently nothing in the embedding appendix that I could find that even describes calling from wasm into the embedder, and thus nowhere to put a description of what happens when such calls throw.

File: document/js-api/index.bs [hunk 7]
```diff
@@ -1336,18 +1355,31 @@ The <dfn attribute for="Exception">stack</dfn> getter steps are:
...
+    1. Set the current agent's [=associated store=] to |store|.
+    1. Set the current agent's associated [=JavaScript exception tag=] to |tagAddress|.
+    1. return |tagAddress|.
+
+</div>
 
 <div algorithm>
 
-To <dfn for=WebAssembly>throw</dfn> with a [=tag address=] |type|, a matching [=list=] of WebAssembly values |payload|, and an [=externref=] |opaqueData|, perform the following steps:
+To <dfn for=WebAssembly>throw</dfn> with a [=tag address=] |type|, a matching [=list=] of WebAssembly values |payload|, and an [=exception address=] |address|, perform the following steps:
```
Comment (rossberg) [inline document/js-api/index.bs]: This should just take an exception address, since that now encapsulates the other values.

File: document/js-api/index.bs [hunk 8]
```diff
@@ -1076,11 +1085,13 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
...
-                1. Let |type| be the [=JavaScript exception tag=].
-                1. Let |payload| be « ».
-            1. Let |opaqueData| be [=ToWebAssemblyValue=](|v|, [=externref=])
-            1. [=WebAssembly/Throw=] with |type|, |payload| and |opaqueData|.
+                1. Let |type| be the result of [=get the JavaScript exception tag |getting the JavaScript exception tag=].
+                1. Let |payload| be [=ToWebAssemblyValue=](|v|, [=externref=]).
+                1. Let (|store|, |address|) be [=exn_alloc=](|store|, |type|, « |payload| »).
+            1. [=WebAssembly/Throw=] with |address|.
         1. Otherwise, return |result|.\[[Value]].
     1. Let |store| be the [=surrounding agent=]'s [=associated store=].
```
Comment (rossberg) [inline document/js-api/index.bs]: This needs to move up now, since the store is already used above.

File: document/js-api/index.bs [hunk 9]
```diff
@@ -1076,11 +1085,13 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
             1. If |v| [=implements=] {{Exception}},
                 1. Let |type| be |v|.\[[Type]].
                 1. Let |payload| be |v|.\[[Payload]].
```
Comment (rossberg) [inline document/js-api/index.bs]: These are unused.
```suggestion
```

File: document/js-api/index.bs [hunk 10]
```diff
@@ -1076,11 +1085,13 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
...
+                1. Let |address| be |v|.\[[Address]].
+                1. [=WebAssembly/Throw=] with |address|.
```
Comment (rossberg) [inline document/js-api/index.bs]: This is redundant, you can just fall through.
```suggestion
```

File: document/js-api/index.bs [hunk 11]
```diff
@@ -1288,28 +1307,28 @@ lt="Exception(exceptionTag, payload, options)">new Exception(|exceptionTag|, |pa
 constructor steps are:
 
 1. Let |store| be the [=surrounding agent=]'s [=associated store=].
-1. Let |types| be [=tag_parameters=](|store|, |exceptionTag|.\[[Address]]).
+1. Let [|types|] → [] be [=tag_type=](|store|, |exceptionTag|.\[[Address]]).
 1. If |types|'s [=list/size=] is not |payload|'s [=list/size=],
     1. Throw a {{TypeError}}.
 1. Let |wasmPayload| be « ».
 1. [=list/iterate|For each=] |value| and |resultType| of |payload| and |types|, paired linearly,
     1. [=list/Append=] ? [=ToWebAssemblyValue=](|value|, |resultType|) to |wasmPayload|.
```
Comment (rossberg) [inline document/js-api/index.bs]:
```suggestion
    1. [=list/Append=] [=?=] [=ToWebAssemblyValue=](|value|, |resultType|) to |wasmPayload|.
```

File: document/js-api/index.bs [hunk 12]
```diff
@@ -1076,11 +1085,13 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
...
                 1. Let |payload| be |v|.\[[Payload]].
+                1. Let |address| be |v|.\[[Address]].
+                1. [=WebAssembly/Throw=] with |address|.
             1. Otherwise,
-                1. Let |type| be the [=JavaScript exception tag=].
-                1. Let |payload| be « ».
-            1. Let |opaqueData| be [=ToWebAssemblyValue=](|v|, [=externref=])
-            1. [=WebAssembly/Throw=] with |type|, |payload| and |opaqueData|.
+                1. Let |type| be the result of [=get the JavaScript exception tag |getting the JavaScript exception tag=].
+                1. Let |payload| be [=ToWebAssemblyValue=](|v|, [=externref=]).
```
Comment (rossberg) [inline document/js-api/index.bs]: I think there is a meta-level type error here: extrapolating from its internal use of `[=?=]`, ToWebAssemblyValue supposedly returns a JS completion and hence can e.g. fail with an exception, so we must assert here that it is a normal completion. (I think that's what `[=!=]` does in the JS spec.)

And FWIW I just noticed that there is a similar (pre-existing) problem with various other use sites of ToWebAssemblyValue failing to use `[=?=]`.
```suggestion
                1. Let |payload| be [=!=] [=ToWebAssemblyValue=](|v|, [=externref=]).
```

File: document/js-api/index.bs (hunk 6 above)
Comment (rossberg) [reply document/js-api/index.bs]: The problem is that this can not be an "external" function, since it could only be invoked inside a Wasm context. So it can't easily be defined as a proper embedder function, which need a closed evaluation context. To make it work, we would need to pass it the "current evaluation context", which would require means for reifying an entire host/Wasm execution. There is no way to make that work with the JS spec style.

But this algorithm is invoked from another one that already describes Wasm semantics, not host semantics. So we're morally "inside Wasm". So what I'd suggest is to say something like "execute the Wasm instructions `(ref.exn |address|) (throw_ref)`.

And with my other comment there is only one place left where this is invoked, so I'd just inline that step there and remove this algorithm.

File: document/js-api/index.bs (hunk 12 above)
Comment (dschuff) [reply document/js-api/index.bs]: Yeah that makes sense, and you are right about `!`:
https://tc39.es/ecma262/#sec-returnifabrupt-shorthands

Comment (dschuff) [reply document/js-api/index.bs]: Probably makes sense to fix the other uses of ToWebAssemblyValue in a separate PR.

File: document/js-api/index.bs (hunk 6 above)
Comment (dschuff) [reply document/js-api/index.bs]: Makes sense, done.

**Comment by dschuff** [comment]
> But this algorithm is invoked from another one that already describes Wasm semantics, not host semantics. So we're morally "inside Wasm". So what I'd suggest is to say something like "execute the Wasm instructions (ref.exn |address|) (throw_ref).
> And with my other comment there is only one place left where this is invoked, so I'd just inline that step there and remove this algorithm.

This is done too, but since the lines it was attached to are now gone, github won't let me reply...

File: document/js-api/index.bs [hunk 13]
```diff
@@ -1336,20 +1355,22 @@ The <dfn attribute for="Exception">stack</dfn> getter steps are:
...
-1. Invoke the catch block with |payload| and |opaqueData|.
+To <dfn>get the JavaScript exception tag</dfn>, perform the following steps:
 
-Note: This algorithm is expected to be moved into the core specification.
+    1. If the [=surrounding agent=]'s associated [=JavaScript exception tag=] has been initialized,
+        1. return the [=surrounding agent=]'s associated [=JavaScript exception tag=]
+    1. Let |store| be the [=surrounding agent=]'s [=associated store=].
+    1. Let (|store|, |tagAddress|) be [=tag_alloc=](|store|, « [=externref=] » → « »).
+    1. Set the current agent's [=associated store=] to |store|.
+    1. Set the current agent's associated [=JavaScript exception tag=] to |tagAddress|.
```
Comment (aheejin) [inline document/js-api/index.bs:1372]: Does this mean tag == tagAddress? Somewhere above in this doc contains
```
Set map[tagAddress] to tag.
```
So they seem to be different.

And this PR doesn't include how Wasm should handle the JS exception tag, right?

File: document/js-api/index.bs [hunk 14]
```diff
@@ -1071,18 +1082,18 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
         1. [=Clean up after running a callback=] with |stored settings|.
         1. [=Clean up after running script=] with |relevant settings|.
         1. Assert: |result|.\[[Type]] is <emu-const>throw</emu-const> or <emu-const>normal</emu-const>.
+        1. Let |store| be the [=surrounding agent=]'s [=associated store=].
         1. If |result|.\[[Type]] is <emu-const>throw</emu-const>, then:
             1. Let |v| be |result|.\[[Value]].
             1. If |v| [=implements=] {{Exception}},
-                1. Let |type| be |v|.\[[Type]].
-                1. Let |payload| be |v|.\[[Payload]].
+                1. Let |address| be |v|.\[[Address]].
```
Comment (aheejin) [inline document/js-api/index.bs:1089]: Does a `WebAssembly.Exception` object have `Address`?

File: document/js-api/index.bs [hunk 15]
```diff
@@ -1071,18 +1082,18 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
...
             1. Otherwise,
-                1. Let |type| be the [=JavaScript exception tag=].
-                1. Let |payload| be « ».
-            1. Let |opaqueData| be [=ToWebAssemblyValue=](|v|, [=externref=])
-            1. [=WebAssembly/Throw=] with |type|, |payload| and |opaqueData|.
+                1. Let |type| be the result of [=get the JavaScript exception tag |getting the JavaScript exception tag=].
+                1. Let |payload| be [=!=] [=ToWebAssemblyValue=](|v|, [=externref=]).
```
Comment (aheejin) [inline document/js-api/index.bs:1092]: It looks `v` contains `Type` and `Value`. Shouldn't we only convert the `Value` part, rather than the whole `v`?

File: document/js-api/index.bs [hunk 16]
```diff
@@ -1071,18 +1082,18 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
...
+                1. Let (|store|, |address|) be [=exn_alloc=](|store|, |type|, « |payload| »).
+                1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
+            1. Execute the wasm instructions ([=ref.exn=] |address|) ([=throw_ref=])
```
Comment (rossberg) [inline document/js-api/index.bs]:
```suggestion
            1. Execute the WebAssembly instructions ([=ref.exn=] |address|) ([=throw_ref=]).
```

**Review by rossberg — APPROVED** [review]

File: document/js-api/index.bs (hunk 14 above)
Comment (dschuff) [reply document/js-api/index.bs:1089]: Yes, `Address` is initialized in "initialize an Exception object". Exception is now just like all the other JS-exposed objects which also have their wasm Addresses stored in an internal slot in the JS object. (Although I just noticed that the naming convention isn't consistent right now; The `Exception` and `Tag` objects have the address in a field named `Address` whereas the rest have that field named after the object type, e.g. `Global`. Maybe I'll fix that in a followup change).

File: document/js-api/index.bs (hunk 15 above)
Comment (dschuff) [reply document/js-api/index.bs:1092]: The `Type` and `Value` fields come from [ECMAScript completion record](https://tc39.es/ecma262/#sec-completion-record-specification-type), and `result` is the completion record.
On lines 1086 - 1087 above, `result.Type` is `THROW` and then  `v` is `result.Value`, i.e. the value returned by the JS host function called on line 1081, so it's already a value. So I think that's the thing that we need to convert to a wasm externref and throw back into wasm.

... actually though, I just noticed that the handling of completion records between 'run a host function' and 'create a host function' wasn't quite right even before the EH spec. I just filed https://github.com/WebAssembly/spec/issues/1743 about that.

File: document/js-api/index.bs (hunk 13 above)
Comment (dschuff) [reply document/js-api/index.bs:1372]: Line 1208 is where you're referring to. There `tag` and `tagAddress` are both basically function parameters, and `tag` is a JS object and `tagAddress` is an address as defined in the wasm runtime spec. Here `tagAddress` is basically a local variable (a wasm address), a return value of `tag_alloc`; and `tag` isn't a variable at all. This function just returns the address.

And yes,  #269 is the PR that allows actually exposing the JS exception tag (which by extension allows it to be imported into a wasm module).

File: document/js-api/index.bs [hunk 17]
```diff
@@ -1011,15 +1021,16 @@ This slot holds a [=function address=] relative to the [=surrounding agent=]'s [
...
-        1. If |opaqueData| is not [=ref.null=] [=externref=],
-            1. Let « [=ref.extern=] |externaddr| » be |opaqueData|.
-            1. Throw the result of [=retrieving an extern value=] from |externaddr|.
-        1. Let |exception| be [=create an Exception object|a new Exception=] for |exntag| and |payload|.
-        1. Throw |exception|.
+    1. If |ret| is [=THROW=] [=ref.exn=] |exnaddr|, then
+        1. Let (|tagaddr|, |payload|) be [=exn_read=](|store|, |exnaddr|).
+        1. Let |jsTagAddr| be the result of [=get the JavaScript exception tag |getting the JavaScript exception tag=].
+        1. If |tagaddr| is equal to |jsTagAddr|,
+            1. Throw the result of [=retrieving an extern value=] from |payload|[0].
```
Comment (thibaudmichaud) [inline document/js-api/index.bs:1030]: There is an optimization that was discussed before where the engine does not wrap/unwrap the exception at the boundary, and instead implements the JSTag behavior as a special-case in the catch handler. But with that optimization, if the user code creates a `WebAssembly.Exception(WebAssembly.JSTag, ...)` object explicitly and throws it, it would not get unwrapped automatically at the export, which would violate this spec, right?
We can still benefit from the optimization and add a check when the object crosses the JS/Wasm boundary to cover this special case, but I just wanted to make sure that this is indeed what this spec requires us to do.

Comment (sjrd) [reply document/js-api/index.bs:1030]: I was thinking about that issue as well while porting the `JSTag` behavior to the `exnref` implementation in V8. I was wondering if we could disallow creating a `WebAssembly.Exception(JSTag, ...)` in the first place? As in, throw a `RangeError` or some other exception from the constructor of `WA.Exception`. That would probably simplify a number of things, both at the spec and implementation level.

Would that behavior be acceptable?

Comment (dschuff) [reply document/js-api/index.bs:1030]: I would describe the spec as currently written as saying that regular JS exceptions just never get wrapped or unwrapped, the payload just passes through (although like all exceptions the payload is paired with a tag, even if that tag is not exposed to the user).  I think as written today, if you created a `WebAssembly.Exception(WebAssembly.JSTag, payload)` and threw it into wasm, it would be treated as a wasm exception on entry and unrapped; and then when it came back out to JS it would not be rewrapped but come out as just `payload`. If I'm understanding the optimization you describe, it would naturally have exactly this behavior, so there wouldn't need to be special cases.

Having said that, that behavior is a little weird and may be unexpected since it's asymmetric. I don't really see a use case for an explicitly-wrapped `WebAssembly.Exception` with the JS tag, so maybe it does make sense to disallow creation of such an object.

Comment (dschuff) [reply document/js-api/index.bs:1030]: I also realized that this discussion should probably go in #269 since that PR is where the JS Tag object actually gets exposed to JS explicitly and makes that behavior possible.

**Comment by dschuff** [comment]
I'm going to go ahead and land this so I can rebase #269 and do some more cleanups. @aheejin let me know if you have more comments, we can update as needed.
